lib/network/modules/components/nodes/util/ShapeBase.js

```diff
@@ -43,16 +43,21 @@ class ShapeBase extends NodeBase {
    * @param {boolean} selected
    * @param {boolean} hover
    * @param {ArrowOptions} values
+   * @param {function} customRenderer - a custom shape renderer similar to getShape(shape) functions
    * @private
    */
-  _drawShape(ctx, shape, sizeMultiplier, x, y, selected, hover, values) {
+  _drawShape(ctx, shape, sizeMultiplier, x, y, selected, hover, values, customRenderer) {
     this.resize(ctx, selected, hover, values);
     this.left = x - this.width / 2;
     this.top = y - this.height / 2;
 
     this.initContextForDraw(ctx, values);
-    getShape(shape)(ctx, x, y, values.size);
-    this.performFill(ctx, values);
+    if (shape === 'custom') {
+      customRenderer(ctx, x, y, selected, hover, values)
```

Comment:
We should probably `ctx.save()` and `ctx.restore()` around this. Who knows what they'll set and won't clear in this function.

Reply:
Agreed! good catch.